core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -554,25 +477,34 @@ private object RTS {
                     curIo = enterSupervision *>
                       io.value.ensuring(exitSupervision(io.supervisor))
 
-                  case IO.Tags.Terminate =>
-                    val io = curIo.asInstanceOf[IO.Terminate]
+                  case IO.Tags.Fail =>
+                    val io = curIo.asInstanceOf[IO.Fail[E]]
 
-                    val finalizer = interruptStack
+                    val finalizer = unwindStack(io.cause.isChecked)
 
-                    if (finalizer eq null) {
-                      // No finalizers, simply produce error:
-                      val causes    = status.get.terminationCauses.getOrElse(Nil)
-                      val defects   = status.get.defects
-                      val allCauses = io.causes ++ causes ++ defects
+                    if (stack.isEmpty) {
+                      // Error not caught, stack is empty:
+                      if (finalizer eq null) {
+                        // No finalizer, so immediately produce the error.
+                        curIo = null
 
-                      curIo = null
-                      result = ExitResult.Terminated(allCauses)
+                        val cause = if (status.get.interrupted) io.cause ++ Cause.interrupted else io.cause
+                        result = ExitResult.failed(cause)
+                      } else {
+                        // We have finalizers to run. We'll resume executing with the
+                        // uncaught failure after we have executed all the finalizers:
+                        val completer = io
```

Comment:
Because `io` is already a local val, you can inline it into the subsequent line, rather than using `completer`.